text/0000-ppv.md

```diff
@@ -0,0 +1,1441 @@
+/// Are "all" lanes `true`?
+pub fn all(self) -> bool;
+
+/// Is "any" lane `true`?
+pub fn any(self) -> bool;
+
+/// Are "all" lanes `false`?
+pub fn none(self) -> bool;
+}
+```
+
+#### Mask vertical selection
+
+All vector masks implement the following method:
+
+```rust
+impl m{lane_width}x{number_of_lanes} {
+/// Lane-wise selection. 
+///
+/// The lanes of the result for which the mask is `true` contain
+/// the values of `a` while the remaining lanes contain the values of `b`.
+pub fn select<T>(self, a: T, b: T) -> T
+    where T::lanes() == number_of_lanes; // implementation-defined 
+}
+```
+
+Note: how `where` clause is enforced is an implementation detail. `stdsimd`
+implements this using a sealed trait:
+
+```rust
+pub fn select<T>(self, a: T, b: T) -> T
+    where T: SelectMask<Self>
+}
+```
+
+#### Vertical comparisions
+
+All vector types implement the following vertical (lane-wise) comparison methods
+that returns a mask expressing the result.
+
+```rust
+impl {element_type}{lane_width}x{number_of_lanes} {
+/// Lane-wise equality comparison.
+pub fn eq(self, other: $id) -> m{lane_width}x{number_of_lanes};
+
+/// Lane-wise inequality comparison.
+pub fn ne(self, other: $id) -> m{lane_width}x{number_of_lanes};
+
+/// Lane-wise less-than comparison.
+pub fn lt(self, other: $id) -> m{lane_width}x{number_of_lanes};
+
+/// Lane-wise less-than-or-equals comparison.
+pub fn le(self, other: $id) -> m{lane_width}x{number_of_lanes};
+
+/// Lane-wise greater-than comparison.
+pub fn gt(self, other: $id) -> m{lane_width}x{number_of_lanes};
+
+/// Lane-wise greater-than-or-equals comparison.
+pub fn ge(self, other: $id) -> m{lane_width}x{number_of_lanes};
+}
+```
+
+For all vector types proposed in this RFC, the `{lane_width}` of the mask
+matches that of the vector type. However, this will not be the case for the
+AVX-512 vector types.
+
+##### Semantics for floating-point numbers
+
+* `eq`: yields `true` if both operands are not a `QNAN` and `self` is equal to
+  `other`, yields `false` otherwise.
+* `gt`: yield `true` if both operands are not a `QNAN` and ``self`` is greater
+  than `other`, yields `false` otherwise.
+* `ge`: yields `true` if both operands are not a `QNAN` and `self` is greater
+  than or equal to `other`, yields `false` otherwise.
+* `lt`: yields `true` if both operands are not a `QNAN` and `self` is less than
+  `other`, yields `false` otherwise.
+* `le`: yields `true` if both operands are not a `QNAN` and `self` is less than
+  or equal to `other`, yields `false` otherwise.
+* `ne`: yields `true` if either operand is a `QNAN` or `self` is not equal to
+  `other`, yields `false` otherwise.
+  
+  
+### Portable vector shuffles
+
+```
+/// Shuffles vector elements.
+std::simd::shuffle!(...);
+```
+
+The `shuffle!` macro returns a new vector that contains a shuffle of the elements in
+one or two input vectors. That is, there are two versions:
+
+ * `shuffle!(vec, [indices...])`: one-vector version
```

Comment:
What form do the indices take? Can they be arbitrary constant expressions, or do they have to be literals? Obviously we want the former but this should be mentioned, since a natural way to implement a *macro* would be to match on literals.

Reply:
> Can they be arbitrary constant expressions, or do they have to be literals?

I'll mention that arbitrary constant expressions are allowed.